### PR TITLE
[gpinitsystem2.0] Support for expansion configuration parameter with gp init cluster

### PR DESCRIPTION
**Pull request by piyushc01**
Currently`gp init cluster` supports only absolute configuration with the config file. When specifying the cluster details, we need to specify all the details of all segments array like port, address, data directory etc. 
When creating large clusters it becomes difficult to create and maintain such configurations. To simplify cluster creation and configuration specification, this PR proposes expansion configuration for `gp init cluster`. 
With expansion configuration, we just need to mention the hostlist, base ports and directories for primaries and mirror (optional). Configuration will be expanded in memory to create the segment array and cluster creation will continue. 

Flow to create the cluster remains the same, and no changes in the existing flags definitions. 
To create cluster use the following command: 
`gp init cluster <config-file.json>`

## Program Flow: 
Command detects if expansion keywords are specified in the configuration and creates segment array. 
Flow for expansion configuration is as follows: 
1. Read the json config file from the userin `init.go:LoadInputConfigToIdlFn`
2.  Checks if any expansion configuration is present : `init.go:AnyExpansionConfigPresent()`
3.  Validates provided expansion configuration and also populates defaults: `init.go:ValidateExpansionConfigAndSetDefault()`
4. Check for Multi-home: `init.go:IsMultiHome()`.  Checks if the given setup is multi-home or not. Makes a call to get hostnames from the hosts by connecting using given address. Checks if multiple addresses from given hostlist are pointing to the same host. 
5. Expand Array: `init.go:ExpandSegPairArray()`  expands given configuration to segments list. Also expands mirror if required. Expansion will change if it is a multi-home environment, also on type of mirroring. 
6. Populate config with newly created `SegmentArray`. 
7. Continue with cluster creation. 

## RPC created as part of this PR: 
1. `Agent: GetHostName()` : Simple RPC to which returns name of the host. This is used in detecting multi-home setup. Same hostname is used to populate in the segments list as host-name against given address. 
2. `Hub:GetAllHostNames()` : Takes hostlist as input and fetches hostname for all the host addresses by calling `GetHostName` on each host. This is used to detect if it is a multi-home environment. 

## Configuration parameters:
Details for the supported expansion keywords are as follows:
| Index | Parameter Name  | Purpose | Default Value | 
| ----- | ------------- | ------------- | ------------- |
|  1  | `primary-base-port` | Integer value to specify starting port number for primary segments. he first primary segment port on a host is set as `primary-base-port` , and then incremented by one for each additional primary segment on that host.|  Coordinator Port + 2 |
| 2   | `primary-data-directories`  | Array of strings to specify base path for data directories for primary segments on the host. Number of entries in the array defines number of segments per host. If want to specify different path for each segments, specify the list accordingly  | Mandatory parameter. No default | 
| 3 | `hostlist` | Array of strings specifies list of hosts where primaries and mirrors will be created. Can specify host-address/name here and hostname will be resolved from the host using RPC. Value given here will be populated as host-address in the cluster configuration when creating cluster. In case of multi-home environment, specify all the host addresses here as different hosts. Based on hostname, we will check if the setup is multi-home or not. Also make sure SSL certificates have given address specified in the certificate to avoid connection errors. | Mandatory parameter. No default |
| 4 |  `mirror-base-port`  | Integer value to specify starting port number for mirrors. Used only when `mirror-data-directories` is specified. The first mirror segment port on a host is set as `mirror-base-port` , and then incremented by one for each additional mirror segment on that host.  |  primary-base-port + 1000 | 
| 5 |  `mirror-data-directories` |  Array of strings to specify base path for data directories for mirror segments on the host. Number of entries in the array must be equal to entries in `primary-data-directories` list. If want to specify different path for each segments, specify the list accordingly. | Mandatory parameter. No default if want to specify mirrors. | 
| 6 | `mirroring-type` |  Specifies type of mirroring configuration. Can specify group and spread mirroring. In group mirroring groups the mirror segments for all of a host's primary segments on a single alternate host. Spread mirroring, spreads mirror segments for the primary segments on a host across different hosts in the hostlist. Number of hosts should be more than segments per host to enable spread mirroring | Default is `group` mirorring only when mirrors are specified  | 

## Testing Done: 
- Added unit tests for most of the functions. ~80% tests coverage. 
- Created pipelines, ran lint and go format tool. 
- Manually tested cluster creation from given expansion configuration

## Sample configurations:
Sample configuration file for a demo cluster (only primaries): 
```
{
	"db-name":"gpdb-demo",
	"coordinator":{
		"hostname":"cdw",
		"address":"cdw",
		"port":7000,
		"data-directory":"/tmp/demo/coordinator"
	},
	"primary-base-port":7002,
	"primary-data-directories": [ "/tmp/demo/", "/tmp/demo", "/tmp/demo", "/tmp/demo" ],
	"hostlist": [ "cdw" ]
}
```

Sample configuration file for a demo cluster with mirrors: 
```
{
	"db-name":"gpdb-demo",
	"coordinator":{
		"hostname":"cdw",
		"address":"cdw",
		"port":7000,
		"data-directory":"/tmp/demo/coordinator"
	},
	"primary-base-port":7002,
	"primary-data-directories":[ "/tmp/demo", "/tmp/demo", "/tmp/demo", "/tmp/demo" ],
	"mirror-base-port":8002,
	"mirror-data-directories":[ "/tmp/demo/mirror", "/tmp/demo/mirror", "/tmp/demo/mirror", "/tmp/demo/mirror" ],
	"mirroring-type":"group",
	"hostlist":[ "cdw" ]
}
```

Sample configuration when mounting multiple segments on a host  to different mount points/directories: 
```
{
	"db-name":"gpdb-demo",
	"coordinator":{
		"hostname":"cdw",
		"address":"cdw",
		"port":7000,
		"data-directory":"/tmp/demo/coordinator"
	},
	"primary-base-port":7002,
	"primary-data-directories":[ "/data1", "/data2", "/data3", "/data4" ],
	"mirror-base-port":8002,
	"mirror-data-directories":[ "/data1/mirror", "/data2/mirror", "/data3/mirror", "/data4/mirror" ],
	"mirroring-type":"group",
	"hostlist":[ "cdw" ]
}
```

Sample configuration file for a multi host configuration:
```
{
	"db-name":"gpdb-multihost",
	"coordinator":{
		"hostname":"cdw",
		"address":"cdw",
		"port":7000,
		"data-directory":"/tmp/demo/coordinator"
	},
	"primary-base-port":7002,
	"primary-data-directories":[ "/tmp/demo", "/tmp/demo", "/tmp/demo", "/tmp/demo" ],
	"mirror-base-port":8002,
	"mirror-data-directories":[ "/tmp/demo/mirror", "/tmp/demo/mirror", "/tmp/demo/mirror", "/tmp/demo/mirror" ],
	"mirroring-type":"group",
	"hostlist":[ "sdw1" , "sdw2", "sdw3", "sdw4" ]
}
```

Multihost multi-home configuration when each host has 2 interfaces (e.g. sdw1 has 2 interfaces sdw1-1 ans sdw1-2): 
```
{
	"db-name":"gpdb-multihome",
	"coordinator":{
		"hostname":"cdw",
		"address":"cdw",
		"port":7000,
		"data-directory":"/tmp/demo/coordinator"
	},
	"primary-base-port":7002,
	"primary-data-directories":[ "/tmp/demo", "/tmp/demo", "/tmp/demo", "/tmp/demo" ],
	"mirror-base-port":8002,
	"mirror-data-directories":[ "/tmp/demo/mirror", "/tmp/demo/mirror" ],
	"mirroring-type":"group",
	"hostlist":[ "sdw1-1" , "sdw1-2",  "sdw2-1", "sdw2-2", "sdw3-1", "sdw3-2" ]
}
```

### CODE REVIEW

File: gpMgmt/bin/go-tools/cli/init.go [hunk 1]
```diff
@@ -171,9 +183,346 @@ func LoadInputConfigToIdlFn(inputConfigFile string, cliHandler *viper.Viper, for
...
+			// Default is group mirroring
+			config.MirroringType = constants.GroupMirroringStr
+			gplog.Warn("Mirroring type not specified. Setting default as 'group' mirroring")
+		} else if strings.ToLower(config.MirroringType) != constants.SpreadMirroringStr && strings.ToLower(config.MirroringType) != constants.GroupMirroringStr {
+			strErr := fmt.Sprintf("mirroring-Type: %s is not supported. Only 'group' or 'spread' mirroring is supported",
+				config.MirroringType)
+			gplog.Error(strErr)
+			return fmt.Errorf(strErr)
+		}
+		config.MirroringType = strings.ToLower(config.MirroringType)
```
Comment (jnihal) [inline gpMgmt/bin/go-tools/cli/init.go:336]: We can make this conversion before itself and then perform the validation. This way we do not need to call `strings.ToLower` multiple times

File: gpMgmt/bin/go-tools/cli/init.go [hunk 2]
```diff
@@ -171,9 +183,346 @@ func LoadInputConfigToIdlFn(inputConfigFile string, cliHandler *viper.Viper, for
...
+			config.MirrorBasePort = defaultMirrorBasePort
+		}
+		// Check if spread mirroring, number of hosts should be greater than number of primaries per host so that we can spread segments
+		if config.MirroringType == constants.SpreadMirroringStr && len(config.MirrorDataDirectories) > len(config.HostList) {
+			strErr := fmt.Sprintf("To enable spread mirroring, number of hosts should be more than number of primary segments per host. "+
+				"Current number of hosts is: %d and number of primaries per host is:%d", len(config.HostList), len(config.MirrorDataDirectories))
+			gplog.Error(strErr)
+			return fmt.Errorf(strErr)
+		}
+		//TODO Check if the primary and mirror range overlaps
```
Comment (jnihal) [inline gpMgmt/bin/go-tools/cli/init.go:324]: Is this still valid?

File: gpMgmt/bin/go-tools/cli/init.go [hunk 3]
```diff
@@ -171,9 +183,346 @@ func LoadInputConfigToIdlFn(inputConfigFile string, cliHandler *viper.Viper, for
...
+	} else {
+		// Mirror-less configuration
+		strErr := "No mirror-data-directories provided. Will create mirror-less cluster"
+		gplog.Warn(strErr)
+		ContainsMirror = false
+	}
+
+	// If provided expansion config and primary/mirrors array is also defined
+	if len(config.SegmentArray) > 0 {
+		strErr := "segments-array list should be empty when configuration contains primary-base-directories and hostlist"
```
Comment (jnihal) [inline gpMgmt/bin/go-tools/cli/init.go:346]: This check should be performed at the beginning of this function and better to use `viper.IsSet`.
Also `segments-array list should be empty when configuration contains primary-base-directories and hostlist` -> `cannot specify segments-array and primary-base-directories together`

File: gpMgmt/bin/go-tools/cli/init.go [hunk 4]
```diff
@@ -171,9 +183,346 @@ func LoadInputConfigToIdlFn(inputConfigFile string, cliHandler *viper.Viper, for
...
+		gplog.Error(strErr)
+		return fmt.Errorf(strErr)
+	}
+
+	// TODO Check related to multi-homing
```
Comment (jnihal) [inline gpMgmt/bin/go-tools/cli/init.go:351]: Is this still valid?

File: gpMgmt/bin/go-tools/cli/init.go [hunk 5]
```diff
@@ -171,9 +183,346 @@ func LoadInputConfigToIdlFn(inputConfigFile string, cliHandler *viper.Viper, for
...
+
+func ExpandNonMultiHomePrimaryList(segPairList *[]SegmentPair, PrimaryBasePort int, PrimaryDataDirectories []string, hostList []string, addressNameMap map[string]string) *[]SegmentPair {
+	segNum := 0
+	for _, hostAddress := range hostList {
+		for segIdx, directory := range PrimaryDataDirectories {
+			seg := Segment{
+				Hostname:      addressNameMap[hostAddress],
+				Address:       hostAddress,
+				Port:          PrimaryBasePort + segIdx,
+				DataDirectory: filepath.Join(directory, fmt.Sprintf("gpseg-%d", segNum)),
```
Comment (jnihal) [inline gpMgmt/bin/go-tools/cli/init.go:363]: We should not be adding `-` to the directories as the coordinator value will be `gpseg-1`.
Also `gpseg` should be defined as a constant.

File: gpMgmt/bin/go-tools/constants/constants.go [hunk 6]
```diff
@@ -18,4 +18,6 @@ const (
 	DefaultEncoding     = "UTF-8"
 	RolePrimary         = "p"
 	EtcHostsFilepath    = "/etc/hosts"
+	GroupMirroringStr   = "group"
+	SpreadMirroringStr  = "spread"
```
Comment (jnihal) [inline gpMgmt/bin/go-tools/constants/constants.go:22]: `GroupMirroringStr` -> `GroupMirroring`
`SpreadMirroringStr` -> `SpreadMirroring`

**Review by jnihal — COMMENTED** [review]
First pass: Mostly looked through the CLI side of things. Some things to consider throughout the code-base
- Add newlines wherever required (between functions, before a return statement, multiple if-else blocks) to better distinguish the code blocks.
- Add a space after the colon `:` while formatting strings. This would make the output more readable to the user.

File: gpMgmt/bin/go-tools/cli/init.go (hunk 2 above)
Comment (piyushc01) [reply gpMgmt/bin/go-tools/cli/init.go:324]: Yes, this check needs to be incorporated in future

File: gpMgmt/bin/go-tools/cli/init.go (hunk 4 above)
Comment (piyushc01) [reply gpMgmt/bin/go-tools/cli/init.go:351]: Covered as part of multi home checks, marking resolved

File: gpMgmt/bin/go-tools/cli/init.go [hunk 7]
```diff
@@ -171,9 +183,346 @@ func LoadInputConfigToIdlFn(inputConfigFile string, cliHandler *viper.Viper, for
...
 
+	if AnyExpansionConfigPresent(cliHandler) {
+		// Validate expansion config
+		err := ValidateExpansionConfigAndSetDefault(&config, cliHandler)
+		if err != nil {
+			return &idl.MakeClusterRequest{}, err
+		}
+
+		// Check for multi-home
+		isMultHome, NameAddressMap, AddressNameMap, err := IsMultiHome(config.HostList)
```
Comment (shrakesh) [inline gpMgmt/bin/go-tools/cli/init.go:194]: nit: isMultHome -> isMultiHome

File: gpMgmt/bin/go-tools/cli/init.go [hunk 8]
```diff
@@ -171,9 +183,346 @@ func LoadInputConfigToIdlFn(inputConfigFile string, cliHandler *viper.Viper, for
...
+		if err != nil {
+			gplog.Error("multihome detection failed, error:%v", err)
+			return &idl.MakeClusterRequest{}, err
+		}
+		if isMultHome {
+			isValidMutliHomeConfig, err := ValidateMultiHomeConfig(config, NameAddressMap)
```
Comment (shrakesh) [inline gpMgmt/bin/go-tools/cli/init.go:200]: nit: isValidMutliHomeConfig-> isValidMultiHomeConfig

File: gpMgmt/bin/go-tools/cli/init.go [hunk 9]
```diff
@@ -171,9 +183,353 @@ func LoadInputConfigToIdlFn(inputConfigFile string, cliHandler *viper.Viper, for
...
+	// Check if mandatory mirror expansion parameters are provided
+	if AnyExpansionMirrorConfigPresent(cliHandle) {
+		ContainsMirror = true
+		if len(config.PrimaryDataDirectories) != len(config.MirrorDataDirectories) {
+			strErr := "number of primary-data-directories should be equal to number of mirror-data-directories"
+			return fmt.Errorf(strErr)
+		}
+		// validate mirror base port
+		if !cliHandle.IsSet("mirror-base-port") {
+			defaultMirrorBasePort := config.Coordinator.Port + 1002
```
Comment (shrakesh) [inline gpMgmt/bin/go-tools/cli/init.go:325]: There was a recent issue came from the customer that the mirror port should be calculated from mirror base port  + 1000(default offset). it should not be from the coordinator.

either we can consider that now or we can patch it later once the customer jira gets resolved. i will leave it up to you.

Comment (piyushc01) [reply gpMgmt/bin/go-tools/cli/init.go]: Updated to use default mirror-base-port value as `PrimaryBasePort+1000`.

**Review by jnihal — APPROVED** [review]

**Review by shrakesh — APPROVED** [review]
LGTM 
Approved.

**Comment by kmacoskey** [comment]
Just a small comment from ShipIt. 

We've always controlled the number of Segments per Segment Host by the count of data directories. This has always been awkward. It's not often that the directory used on each Segment Host is different. They are often the same. 

So a possible improvement is to only have the user specify a single directory then give a integer for the number of Segments per Segment Host. This could be the default. 

Optionally, let the user still specify specific data directory to use per Segment Host.

**Pull request merged**